server.go

```diff
@@ -303,14 +304,18 @@ func (sp *serverPeer) newestBlock() (*chainhash.Hash, int32, error) {
 // addKnownAddresses adds the given addresses to the set of known addresses to
 // the peer to prevent sending duplicate addresses.
 func (sp *serverPeer) addKnownAddresses(addresses []*wire.NetAddress) {
+	sp.addrMtx.Lock()
+	defer sp.addrMtx.Unlock()
 	for _, na := range addresses {
 		sp.knownAddresses[addrmgr.NetAddressKey(na)] = struct{}{}
 	}
 }
 
 // addressKnown true if the given address is already known to the peer.
 func (sp *serverPeer) addressKnown(na *wire.NetAddress) bool {
+	sp.addrMtx.Lock()
 	_, exists := sp.knownAddresses[addrmgr.NetAddressKey(na)]
+	sp.addrMtx.Unlock()
```

Comment:
NetAddressKey should not have an error in the current implementation but wouldn't it be safer to avoid a deadlock to defer the Unlock here?